### PR TITLE
[RFE] add systemd support

### PR DESCRIPTION
**Pull request by 257**
motivation for this is is proper ordering of related/bound/required services to Hyprland (e.g. swaybg) that would need to have a compositor ready.

this could possibly be a build-time option of course.

see also:
example/ files for example of services

### CODE REVIEW

File: src/Compositor.cpp [hunk 1]
```diff
@@ -368,6 +369,9 @@ void CCompositor::startCompositor() {
 
     wlr_xcursor_manager_set_cursor_image(m_sWLRXCursorMgr, "left_ptr", m_sWLRCursor);
 
+    // tell systemd we're that we're ready so i can start bond/following units
```
Comment (NotAShelf) [inline src/Compositor.cpp:372]: "tell systemd we're that we're ready"

**Comment by vaxerski** [comment]
Not everyone uses systemd. We should make it only use systemd when it's found.

**Comment by 257** [comment]
> Not everyone uses systemd. We should make it only use systemd when it's found.

will add `sd_boot()`, that's runtime though. i will have a look into the build also.

**Comment by 257** [comment]
> define `USES_SYSTEMD` or something in meson & cmake if systemd is found. Then use that for `#ifdef`s

just pushed, how does it look like?

**Comment by vaxerski** [comment]
you're still making libsystemd required in the Makefile + you didnt wrap headers with the guard

**Comment by 257** [comment]
oops! should be fixed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -74,6 +74,16 @@ ELSE()
     target_link_libraries(Hyprland xcb)
 ENDIF(NO_XWAYLAND MATCHES true)
 
+IF(USES_SYSTEMD MATCHES true)
```
Comment (vaxerski) [inline CMakeLists.txt:77]: okay but it still doesnt automatically find it.

Comment (257) [reply CMakeLists.txt:77]: so we want to set `USES_SYSTEMD` unconditionally _if_ `SYSTEMD_FOUND`?

Comment (257) [reply CMakeLists.txt:77]: oh! `cmake` stuff is un-related to `meson`, of course.

File: CMakeLists.txt [hunk 3]
```diff
@@ -54,7 +54,7 @@ message(STATUS "Checking deps...")
 find_package(Threads REQUIRED)
 
 find_package(PkgConfig REQUIRED)
-pkg_check_modules(deps REQUIRED IMPORTED_TARGET wayland-server wayland-client wayland-cursor wayland-protocols cairo pango pangocairo libdrm egl xkbcommon libinput) # we do not check for wlroots, as we provide it ourselves
+pkg_check_modules(deps REQUIRED IMPORTED_TARGET wayland-server wayland-client wayland-cursor wayland-protocols cairo pango pangocairo libdrm libsystemd egl xkbcommon libinput) # we do not check for wlroots, as we provide it ourselves
```
Comment (vaxerski) [inline CMakeLists.txt:57]: now you make libsystemd required.

You need to store whether it was found in a variable, and use it later below

Comment (257) [reply CMakeLists.txt:57]: `cmake` fails here, any idea?

```
foo@host: mkdir build && cd build && cmake .. && make -j
-- The C compiler identification is GNU 12.2.1
-- The CXX compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Gathering git info
-- Configuring Hyprland in Release with CMake
-- Checking deps...
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.0")
-- Checking for modules 'wayland-server;wayland-client;wayland-cursor;wayland-protocols;cairo;libdrm;egl;xkbcommon;libinput'
--   Found wayland-server, version 1.21.0
--   Found wayland-client, version 1.21.0
--   Found wayland-cursor, version 1.21.0
--   Found wayland-protocols, version 1.31
--   Found cairo, version 1.17.6
--   Found libdrm, version 2.4.114
--   Found egl, version 1.5
--   Found xkbcommon, version 1.4.1
--   Found libinput, version 1.22.0
-- XWAYLAND Enabled (NO_XWAYLAND not defined) checking deps...
-- Checking for module 'xcb'
--   Found xcb, version 1.15
-- Setting link libraries
-- Configuring done
-- Generating done
-- Build files have been written to: /home/pmn/src/github/257/Hyprland/build
[  3%] Building CXX object CMakeFiles/Hyprland.dir/src/debug/HyprCtl.cpp.o
[  5%] Building CXX object CMakeFiles/Hyprland.dir/src/debug/HyprDebugOverlay.cpp.o
[  5%] Building CXX object CMakeFiles/Hyprland.dir/src/Compositor.cpp.o
[  7%] Building CXX object CMakeFiles/Hyprland.dir/src/debug/Log.cpp.o
[  9%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Devices.cpp.o
[ 11%] Building CXX object CMakeFiles/Hyprland.dir/src/config/ConfigManager.cpp.o
[ 13%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Layers.cpp.o
[ 15%] Building CXX object CMakeFiles/Hyprland.dir/src/Window.cpp.o
[ 16%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Misc.cpp.o
[ 18%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Monitors.cpp.o
[ 20%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Popups.cpp.o
[ 22%] Building CXX object CMakeFiles/Hyprland.dir/src/events/Windows.cpp.o
[ 24%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/AnimatedVariable.cpp.o
[ 26%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/BezierCurve.cpp.o
[ 28%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/Color.cpp.o
[ 30%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/MiscFunctions.cpp.o
[ 32%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/Monitor.cpp.o
In file included from /home/pmn/src/github/257/Hyprland/src/debug/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/debug/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/debug/../Compositor.hpp:7,
                 from /home/pmn/src/github/257/Hyprland/src/debug/HyprCtl.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/debug/HyprCtl.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[ 33%] Building CXX object CMakeFiles/Hyprland.dir/src/helpers/SubsurfaceTree.cpp.o
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:118: CMakeFiles/Hyprland.dir/src/debug/HyprCtl.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/pmn/src/github/257/Hyprland/src/debug/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/debug/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/debug/Log.cpp:2:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:146: CMakeFiles/Hyprland.dir/src/debug/Log.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/../Compositor.hpp:7,
                 from /home/pmn/src/github/257/Hyprland/src/events/Layers.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/Events.hpp:2,
                 from /home/pmn/src/github/257/Hyprland/src/events/Misc.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:174: CMakeFiles/Hyprland.dir/src/events/Layers.cpp.o] Error 1
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:188: CMakeFiles/Hyprland.dir/src/events/Misc.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/config/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/config/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/config/ConfigManager.hpp:8,
                 from /home/pmn/src/github/257/Hyprland/src/config/ConfigManager.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:104: CMakeFiles/Hyprland.dir/src/config/ConfigManager.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/Window.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/Window.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:90: CMakeFiles/Hyprland.dir/src/Window.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/Events.hpp:2,
                 from /home/pmn/src/github/257/Hyprland/src/events/Popups.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:216: CMakeFiles/Hyprland.dir/src/events/Popups.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/AnimatedVariable.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/AnimatedVariable.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:244: CMakeFiles/Hyprland.dir/src/helpers/AnimatedVariable.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/Compositor.hpp:7,
                 from /home/pmn/src/github/257/Hyprland/src/Compositor.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:76: CMakeFiles/Hyprland.dir/src/Compositor.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/Events.hpp:2,
                 from /home/pmn/src/github/257/Hyprland/src/events/Devices.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:160: CMakeFiles/Hyprland.dir/src/events/Devices.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/debug/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/debug/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/debug/HyprDebugOverlay.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/debug/HyprDebugOverlay.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:132: CMakeFiles/Hyprland.dir/src/debug/HyprDebugOverlay.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/BezierCurve.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/BezierCurve.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:258: CMakeFiles/Hyprland.dir/src/helpers/BezierCurve.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/../Compositor.hpp:7,
                 from /home/pmn/src/github/257/Hyprland/src/events/Monitors.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:202: CMakeFiles/Hyprland.dir/src/events/Monitors.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/events/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/events/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/events/Events.hpp:2,
                 from /home/pmn/src/github/257/Hyprland/src/events/Windows.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:230: CMakeFiles/Hyprland.dir/src/events/Windows.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/Color.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/Color.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:272: CMakeFiles/Hyprland.dir/src/helpers/Color.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/SubsurfaceTree.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/SubsurfaceTree.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/MiscFunctions.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/MiscFunctions.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:286: CMakeFiles/Hyprland.dir/src/helpers/MiscFunctions.cpp.o] Error 1
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:314: CMakeFiles/Hyprland.dir/src/helpers/SubsurfaceTree.cpp.o] Error 1
In file included from /home/pmn/src/github/257/Hyprland/src/helpers/../includes.hpp:53,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/../defines.hpp:1,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/Monitor.hpp:3,
                 from /home/pmn/src/github/257/Hyprland/src/helpers/Monitor.cpp:1:
/home/pmn/src/github/257/Hyprland/subprojects/wlroots/include/wlr/types/wlr_layer_shell_v1.h:16:10: fatal error: wlr-layer-shell-unstable-v1-protocol.h: No such file or directory
   16 | #include "wlr-layer-shell-unstable-v1-protocol.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Hyprland.dir/build.make:300: CMakeFiles/Hyprland.dir/src/helpers/Monitor.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/Hyprland.dir/all] Error 2
make: *** [Makefile:111: all] Error 2
```

Comment (vaxerski) [reply CMakeLists.txt:57]: you forgot to `make protocols` as the wiki instructs you to do with manual debug compiles

**Pull request closed without merging**